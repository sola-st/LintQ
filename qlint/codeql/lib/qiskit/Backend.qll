import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.Register
import qiskit.Gate
import qiskit.Circuit

/** Simulator backend. */
class Backend extends DataFlow::CallCfgNode {
  // case: from qiskit import Aer; Aer.get_backend('qasm_simulator')
  Backend() {
    this = API::moduleImport("qiskit").getMember("Aer").getMember("get_backend").getACall() or
    this = API::moduleImport("qiskit").getMember("BasicAer").getMember("get_backend").getACall()
  }

  /** The name of the backend (according to the initialization string). */
  string getKind() {
    // case: from qiskit import Aer; Aer.get_backend('qasm_simulator')
    // > 'qasm_simulator'
    exists(StrConst str, DataFlow::LocalSourceNode strSource |
      strSource = this.getArg(0) and
      strSource.asExpr() = str
    |
      result = str.getText()
    )
  }

  /** If the simulator is statevector simulator. */
  predicate isStatevectorSimulator() {
    // case: from qiskit import Aer; Aer.get_backend('statevector_simulator')
    this.getKind().matches("%statevector%")
  }

  /** If the simulator is Unitary simulator. */
  predicate isUnitarySimulator() {
    // case: from qiskit import Aer; Aer.get_backend('unitary_simulator')
    this.getKind().matches("%unitary%")
  }

  /** Get a circuit run with this backend. */
  QuantumCircuit getACircuitToBeRun() {
    // case: Aer.get_backend('qasm_simulator').run(circuit)
    // > circuit
    exists(QuantumCircuit circuit, DataFlow::CallCfgNode runCall |
      runCall = this.getAnAttributeRead("run").getACall() and
      (
        circuit.flowsTo(runCall.getArg(0)) or
        circuit.flowsTo(runCall.getArgByName("circuits"))
      )
    |
      result = circuit
    )
    or
    // case: execute(qc, Aer.get_backend('qasm_simulator'), shots=1)
    // > qc
    exists(QuantumCircuit circuit, ExecuteCall executeCall |
      (
        circuit.flowsTo(executeCall.getArg(0)) or
        circuit.flowsTo(executeCall.getArgByName("experiments"))
      ) and
      (
        this.flowsTo(executeCall.getArg(1)) or
        this.flowsTo(executeCall.getArgByName("backend"))
      )
    |
      result = circuit
    )
  }

  /** Get a backend run with this backend. */
  BackendRun getABackendRun() {
    exists(BackendRun bkdRun |
      bkdRun.getBackend() = this
    |
      result = bkdRun
    )
  }

}

/** Backend run.
 *
 * This is a call to a backend run method.
 */
abstract class BackendRun extends DataFlow::CallCfgNode {

  abstract Backend getBackend();

}

/** A call to a backend run method. */
class BackendRunViaRunCall extends BackendRun {
  BackendRunViaRunCall() {
    exists(Backend backend, DataFlow::CallCfgNode runCall |
      runCall = backend.getAnAttributeRead("run").getACall()
    |
      this = runCall
    )
  }

  /** The backend that is run. */
  override Backend getBackend() {
    exists(Backend backend, DataFlow::CallCfgNode runCall |
      runCall = backend.getAnAttributeRead("run").getACall()
    |
      result = backend
    )
  }
}

/** A call to a backend execute method. */
class BackendRunViaExecuteCall extends BackendRun {
  BackendRunViaExecuteCall() {
    this instanceof ExecuteCall
  }

  /** The backend that is run. */
  override Backend getBackend() {
    exists(Backend backend |
      (
        backend.flowsTo(this.getArg(1)) or
        backend.flowsTo(this.getArgByName("backend"))
      )
    |
      result = backend
    )
  }
}

/** Result of a simulator run. */
class BackendResult extends DataFlow::CallCfgNode {
  BackendResult() {
    exists(BackendRun bkdRun |
      this = bkdRun.getAnAttributeRead("result").getACall()
      or
      this = bkdRun.getAnAttributeReference("result")
      or
      this = bkdRun.getAnAttributeWrite("result")
    )
  }

  /** The backend run that generated this result. */
  BackendRun getBackendRun() {
    exists(BackendRun bkdRun |
      // connect the run and its result
      this = bkdRun.getAnAttributeRead("result").getACall() or
      this = bkdRun.getAnAttributeReference("result") or
      this = bkdRun.getAnAttributeWrite("result")
    |
      result = bkdRun
    )
  }

}


/** Statevector returned by a result of a simulator run. */
class Statevector extends DataFlow::CallCfgNode {
  Statevector() {
    exists(BackendResult bkdResult |
      this = bkdResult.getAnAttributeRead("get_statevector").getACall()
    )
  }

  /** The backend run that generated this statevector. */
  BackendRun producedByBackendRun() {
    exists(BackendResult bkdResult, BackendRun bkdRun |
      // connect the run and its result
      bkdResult.getBackendRun() = bkdRun and
      // the result has been called to extract the statevector
      this = bkdResult.getAnAttributeRead("get_statevector").getACall()
    |
      result = bkdRun
    )
  }

}