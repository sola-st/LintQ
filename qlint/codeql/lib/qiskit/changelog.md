2023.03.07
- fix missing gates `crx` and `cry` in the `Gate` class
- improved the `isAppliedAfter` in the `Gate` class to remove false positives when the circuit is re-initialized during the path connecting the two gates.

2023.03.07
- added the getVar() function to the register class to get the variable name of the register.
- added the QubitUsed abstraction to keep track of each access to a qubit, considering also the register it belongs to.

2023.03.03
- refactoring model: java style method calls
- refactoring of register to share part of it.
- removed GateQuantumOperation, substituted with a predicate isMeasurement() on a general gate.

2023.02.28
- refactoring of the Gate class. Remove the tracking of the for loop.

2023.02.21
- improve the efficiency of get_a_target_qubit() of the gate class (remove extra QuantumCircuit variable)
- improved the precision of get_a_target_qubit() of the gate class to account also for qubits which are operated on in a loop (e.g. qc.h(i)).


2023.02.20
- added is_subcircuit() method to Circuit class (by checking if it flows to the compose() or append() method of another circuit object)
- improved get_total_num_qubits() and get_total_num_bits() methods of Circuit class (by considering also the registers added add_register() method)
- in Circuit refactored get_num_qubits_with_integers() and get_num_bits_with_integers() methods to return the static size of the circuit (i.e. QuantumCircuit(2,2) or n = 7; m = 2, QuantumCircuit(n, m)).
- improved get_num_bits() and get_num_qubits() methods of Register class (by considering also local flows of integer variables, e.g. x = 5; QuantumRegister(x))