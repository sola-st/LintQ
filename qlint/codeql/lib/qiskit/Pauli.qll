import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.NoiseModel


class PauliString extends StrConst {
    PauliString() {
        // it can be used in a PauliError
        exists(
            PauliError pauliError
            |
            pauliError.getPauliString() = this
        )
        or
        // it can be used in a Pauli object
        exists(
            DataFlow::CallCfgNode pauliObject
            |
            pauliObject = API::moduleImport("qiskit").getMember("circuit")
                        .getMember("QuantumCircuit").getMember("pauli").getACall()
            or
            // from qiskit.quantum_info import Pauli
            pauliObject = API::moduleImport("qiskit").getMember("quantum_info")
                        .getMember("Pauli").getACall()
            |
            this = pauliObject.getArg(0).asCfgNode().getNode()
        )
    }

    /** Check if the Pauli satisfy the regex ^[+-]?1?[ij]?[IXYZ]+$ */
    predicate isValid() {
        this.getText().regexpMatch("^[+-]?1?[ij]?[IXYZ]+$")
    }

}