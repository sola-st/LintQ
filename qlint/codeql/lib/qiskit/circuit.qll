import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.register


class QuantumCircuit extends DataFlow::CallCfgNode {
    QuantumCircuit() {
        this = API::moduleImport("qiskit").getMember("QuantumCircuit").getACall()
    }

    int get_num_qubits() {
        exists(IntegerLiteral num_qubits |
            num_qubits = this.getArg(0).asExpr() and
            result = num_qubits.getValue()
        ) or
        exists( QuantumRegister qntReg|
            // there is a dataflow between the quantum register and the quantum circuit
            exists(int i |
                qntReg.flowsTo(this.getArg(i)))
            and
            // the number of qubits is the number of qubits in the quantum register
            result = qntReg.get_num_qubits()
        )
    }

    int get_num_bits() {
        exists(IntegerLiteral num_bits |
            num_bits = this.getArg(1).asExpr() and
            result = num_bits.getValue()
        ) or
        exists( ClassicalRegister clsReg|
            // there is a dataflow between the classical register and the quantum circuit
            exists(int i |
                clsReg.flowsTo(this.getArg(i)))
            and
            // the number of bits is the number of bits in the classical register
            result = clsReg.get_num_bits()
        )
    }

    int get_total_num_bits() {
        result = sum(ClassicalRegister clsReg, int i |
            clsReg.flowsTo(this.getArg(i)) |
            clsReg.get_num_bits())
    }

    int get_total_num_qubits() {
        result = sum(QuantumRegister qntReg, int i |
            qntReg.flowsTo(this.getArg(i)) |
            qntReg.get_num_qubits())
    }

}