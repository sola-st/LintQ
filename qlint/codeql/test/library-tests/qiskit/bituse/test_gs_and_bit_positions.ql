import python
import qiskit.BitUse
import qiskit.Circuit

from OperatorSpecification gs
select gs,
  "Operator specification: " + gs.getName() + " - name qubit: " + gs.getAnArgumentNameOfQubit() +
    " - index qubit: " + gs.getAnArgumentIndexOfQubit()
// + " - name param : " + gs.getAnArgumentNameOfParam() + " - index param: " + gs.getAnArgumentIndexOfParam() + " "
