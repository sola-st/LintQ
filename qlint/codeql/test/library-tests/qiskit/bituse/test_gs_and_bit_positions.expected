| CCXGate | Operator specification: CCXGate - name qubit: control_qubit1 - index qubit: 0 |
| CCXGate | Operator specification: CCXGate - name qubit: control_qubit1 - index qubit: 1 |
| CCXGate | Operator specification: CCXGate - name qubit: control_qubit1 - index qubit: 2 |
| CCXGate | Operator specification: CCXGate - name qubit: control_qubit2 - index qubit: 0 |
| CCXGate | Operator specification: CCXGate - name qubit: control_qubit2 - index qubit: 1 |
| CCXGate | Operator specification: CCXGate - name qubit: control_qubit2 - index qubit: 2 |
| CCXGate | Operator specification: CCXGate - name qubit: target_qubit - index qubit: 0 |
| CCXGate | Operator specification: CCXGate - name qubit: target_qubit - index qubit: 1 |
| CCXGate | Operator specification: CCXGate - name qubit: target_qubit - index qubit: 2 |
| CCZGate | Operator specification: CCZGate - name qubit: control_qubit1 - index qubit: 0 |
| CCZGate | Operator specification: CCZGate - name qubit: control_qubit1 - index qubit: 1 |
| CCZGate | Operator specification: CCZGate - name qubit: control_qubit1 - index qubit: 2 |
| CCZGate | Operator specification: CCZGate - name qubit: control_qubit2 - index qubit: 0 |
| CCZGate | Operator specification: CCZGate - name qubit: control_qubit2 - index qubit: 1 |
| CCZGate | Operator specification: CCZGate - name qubit: control_qubit2 - index qubit: 2 |
| CCZGate | Operator specification: CCZGate - name qubit: target_qubit - index qubit: 0 |
| CCZGate | Operator specification: CCZGate - name qubit: target_qubit - index qubit: 1 |
| CCZGate | Operator specification: CCZGate - name qubit: target_qubit - index qubit: 2 |
| CHGate | Operator specification: CHGate - name qubit: control_qubit - index qubit: 0 |
| CHGate | Operator specification: CHGate - name qubit: control_qubit - index qubit: 1 |
| CHGate | Operator specification: CHGate - name qubit: target_qubit - index qubit: 0 |
| CHGate | Operator specification: CHGate - name qubit: target_qubit - index qubit: 1 |
| CPhaseGate | Operator specification: CPhaseGate - name qubit: control_qubit - index qubit: 1 |
| CPhaseGate | Operator specification: CPhaseGate - name qubit: control_qubit - index qubit: 2 |
| CPhaseGate | Operator specification: CPhaseGate - name qubit: target_qubit - index qubit: 1 |
| CPhaseGate | Operator specification: CPhaseGate - name qubit: target_qubit - index qubit: 2 |
| CRXGate | Operator specification: CRXGate - name qubit: control_qubit - index qubit: 1 |
| CRXGate | Operator specification: CRXGate - name qubit: control_qubit - index qubit: 2 |
| CRXGate | Operator specification: CRXGate - name qubit: target_qubit - index qubit: 1 |
| CRXGate | Operator specification: CRXGate - name qubit: target_qubit - index qubit: 2 |
| CRYGate | Operator specification: CRYGate - name qubit: control_qubit - index qubit: 1 |
| CRYGate | Operator specification: CRYGate - name qubit: control_qubit - index qubit: 2 |
| CRYGate | Operator specification: CRYGate - name qubit: target_qubit - index qubit: 1 |
| CRYGate | Operator specification: CRYGate - name qubit: target_qubit - index qubit: 2 |
| CRZGate | Operator specification: CRZGate - name qubit: control_qubit - index qubit: 1 |
| CRZGate | Operator specification: CRZGate - name qubit: control_qubit - index qubit: 2 |
| CRZGate | Operator specification: CRZGate - name qubit: target_qubit - index qubit: 1 |
| CRZGate | Operator specification: CRZGate - name qubit: target_qubit - index qubit: 2 |
| CSGate | Operator specification: CSGate - name qubit: control_qubit - index qubit: 0 |
| CSGate | Operator specification: CSGate - name qubit: control_qubit - index qubit: 1 |
| CSGate | Operator specification: CSGate - name qubit: target_qubit - index qubit: 0 |
| CSGate | Operator specification: CSGate - name qubit: target_qubit - index qubit: 1 |
| CSXGate | Operator specification: CSXGate - name qubit: control_qubit - index qubit: 0 |
| CSXGate | Operator specification: CSXGate - name qubit: control_qubit - index qubit: 1 |
| CSXGate | Operator specification: CSXGate - name qubit: target_qubit - index qubit: 0 |
| CSXGate | Operator specification: CSXGate - name qubit: target_qubit - index qubit: 1 |
| CSdgGate | Operator specification: CSdgGate - name qubit: control_qubit - index qubit: 0 |
| CSdgGate | Operator specification: CSdgGate - name qubit: control_qubit - index qubit: 1 |
| CSdgGate | Operator specification: CSdgGate - name qubit: target_qubit - index qubit: 0 |
| CSdgGate | Operator specification: CSdgGate - name qubit: target_qubit - index qubit: 1 |
| CSwapGate | Operator specification: CSwapGate - name qubit: control_qubit - index qubit: 0 |
| CSwapGate | Operator specification: CSwapGate - name qubit: control_qubit - index qubit: 1 |
| CSwapGate | Operator specification: CSwapGate - name qubit: control_qubit - index qubit: 2 |
| CSwapGate | Operator specification: CSwapGate - name qubit: target_qubit1 - index qubit: 0 |
| CSwapGate | Operator specification: CSwapGate - name qubit: target_qubit1 - index qubit: 1 |
| CSwapGate | Operator specification: CSwapGate - name qubit: target_qubit1 - index qubit: 2 |
| CSwapGate | Operator specification: CSwapGate - name qubit: target_qubit2 - index qubit: 0 |
| CSwapGate | Operator specification: CSwapGate - name qubit: target_qubit2 - index qubit: 1 |
| CSwapGate | Operator specification: CSwapGate - name qubit: target_qubit2 - index qubit: 2 |
| CU1Gate | Operator specification: CU1Gate - name qubit: control_qubit - index qubit: 1 |
| CU1Gate | Operator specification: CU1Gate - name qubit: control_qubit - index qubit: 2 |
| CU1Gate | Operator specification: CU1Gate - name qubit: target_qubit - index qubit: 1 |
| CU1Gate | Operator specification: CU1Gate - name qubit: target_qubit - index qubit: 2 |
| CU3Gate | Operator specification: CU3Gate - name qubit: control_qubit - index qubit: 3 |
| CU3Gate | Operator specification: CU3Gate - name qubit: control_qubit - index qubit: 4 |
| CU3Gate | Operator specification: CU3Gate - name qubit: target_qubit - index qubit: 3 |
| CU3Gate | Operator specification: CU3Gate - name qubit: target_qubit - index qubit: 4 |
| CXGate | Operator specification: CXGate - name qubit: control_qubit - index qubit: 0 |
| CXGate | Operator specification: CXGate - name qubit: control_qubit - index qubit: 1 |
| CXGate | Operator specification: CXGate - name qubit: target_qubit - index qubit: 0 |
| CXGate | Operator specification: CXGate - name qubit: target_qubit - index qubit: 1 |
| CYGate | Operator specification: CYGate - name qubit: control_qubit - index qubit: 0 |
| CYGate | Operator specification: CYGate - name qubit: control_qubit - index qubit: 1 |
| CYGate | Operator specification: CYGate - name qubit: target_qubit - index qubit: 0 |
| CYGate | Operator specification: CYGate - name qubit: target_qubit - index qubit: 1 |
| CZGate | Operator specification: CZGate - name qubit: control_qubit - index qubit: 0 |
| CZGate | Operator specification: CZGate - name qubit: control_qubit - index qubit: 1 |
| CZGate | Operator specification: CZGate - name qubit: target_qubit - index qubit: 0 |
| CZGate | Operator specification: CZGate - name qubit: target_qubit - index qubit: 1 |
| HGate | Operator specification: HGate - name qubit: qubit - index qubit: 0 |
| IGate | Operator specification: IGate - name qubit: qubit - index qubit: 0 |
| Measure | Operator specification: Measure - name qubit: qubit - index qubit: 0 |
| PhaseGate | Operator specification: PhaseGate - name qubit: qubit - index qubit: 1 |
| RVGate | Operator specification: RVGate - name qubit: qubit - index qubit: 3 |
| RXGate | Operator specification: RXGate - name qubit: qubit - index qubit: 1 |
| RXXGate | Operator specification: RXXGate - name qubit: qubit1 - index qubit: 1 |
| RXXGate | Operator specification: RXXGate - name qubit: qubit1 - index qubit: 2 |
| RXXGate | Operator specification: RXXGate - name qubit: qubit2 - index qubit: 1 |
| RXXGate | Operator specification: RXXGate - name qubit: qubit2 - index qubit: 2 |
| RYGate | Operator specification: RYGate - name qubit: qubit - index qubit: 1 |
| RYYGate | Operator specification: RYYGate - name qubit: qubit1 - index qubit: 1 |
| RYYGate | Operator specification: RYYGate - name qubit: qubit1 - index qubit: 2 |
| RYYGate | Operator specification: RYYGate - name qubit: qubit2 - index qubit: 1 |
| RYYGate | Operator specification: RYYGate - name qubit: qubit2 - index qubit: 2 |
| RZGate | Operator specification: RZGate - name qubit: qubit - index qubit: 1 |
| RZXGate | Operator specification: RZXGate - name qubit: qubit1 - index qubit: 1 |
| RZXGate | Operator specification: RZXGate - name qubit: qubit1 - index qubit: 2 |
| RZXGate | Operator specification: RZXGate - name qubit: qubit2 - index qubit: 1 |
| RZXGate | Operator specification: RZXGate - name qubit: qubit2 - index qubit: 2 |
| RZZGate | Operator specification: RZZGate - name qubit: qubit1 - index qubit: 1 |
| RZZGate | Operator specification: RZZGate - name qubit: qubit1 - index qubit: 2 |
| RZZGate | Operator specification: RZZGate - name qubit: qubit2 - index qubit: 1 |
| RZZGate | Operator specification: RZZGate - name qubit: qubit2 - index qubit: 2 |
| Reset | Operator specification: Reset - name qubit: qubit - index qubit: 0 |
| SGate | Operator specification: SGate - name qubit: qubit - index qubit: 0 |
| SXGate | Operator specification: SXGate - name qubit: qubit - index qubit: 0 |
| SdgGate | Operator specification: SdgGate - name qubit: qubit - index qubit: 0 |
| SwapGate | Operator specification: SwapGate - name qubit: qubit1 - index qubit: 0 |
| SwapGate | Operator specification: SwapGate - name qubit: qubit1 - index qubit: 1 |
| SwapGate | Operator specification: SwapGate - name qubit: qubit2 - index qubit: 0 |
| SwapGate | Operator specification: SwapGate - name qubit: qubit2 - index qubit: 1 |
| TGate | Operator specification: TGate - name qubit: qubit - index qubit: 0 |
| TdgGate | Operator specification: TdgGate - name qubit: qubit - index qubit: 0 |
| U1Gate | Operator specification: U1Gate - name qubit: qubit - index qubit: 1 |
| U2Gate | Operator specification: U2Gate - name qubit: qubit - index qubit: 2 |
| U3Gate | Operator specification: U3Gate - name qubit: qubit - index qubit: 3 |
| XGate | Operator specification: XGate - name qubit: qubit - index qubit: 0 |
| YGate | Operator specification: YGate - name qubit: qubit - index qubit: 0 |
| ZGate | Operator specification: ZGate - name qubit: qubit - index qubit: 0 |
| ccx | Operator specification: ccx - name qubit: control_qubit1 - index qubit: 0 |
| ccx | Operator specification: ccx - name qubit: control_qubit1 - index qubit: 1 |
| ccx | Operator specification: ccx - name qubit: control_qubit1 - index qubit: 2 |
| ccx | Operator specification: ccx - name qubit: control_qubit2 - index qubit: 0 |
| ccx | Operator specification: ccx - name qubit: control_qubit2 - index qubit: 1 |
| ccx | Operator specification: ccx - name qubit: control_qubit2 - index qubit: 2 |
| ccx | Operator specification: ccx - name qubit: target_qubit - index qubit: 0 |
| ccx | Operator specification: ccx - name qubit: target_qubit - index qubit: 1 |
| ccx | Operator specification: ccx - name qubit: target_qubit - index qubit: 2 |
| ccz | Operator specification: ccz - name qubit: control_qubit1 - index qubit: 0 |
| ccz | Operator specification: ccz - name qubit: control_qubit1 - index qubit: 1 |
| ccz | Operator specification: ccz - name qubit: control_qubit1 - index qubit: 2 |
| ccz | Operator specification: ccz - name qubit: control_qubit2 - index qubit: 0 |
| ccz | Operator specification: ccz - name qubit: control_qubit2 - index qubit: 1 |
| ccz | Operator specification: ccz - name qubit: control_qubit2 - index qubit: 2 |
| ccz | Operator specification: ccz - name qubit: target_qubit - index qubit: 0 |
| ccz | Operator specification: ccz - name qubit: target_qubit - index qubit: 1 |
| ccz | Operator specification: ccz - name qubit: target_qubit - index qubit: 2 |
| ch | Operator specification: ch - name qubit: control_qubit - index qubit: 0 |
| ch | Operator specification: ch - name qubit: control_qubit - index qubit: 1 |
| ch | Operator specification: ch - name qubit: target_qubit - index qubit: 0 |
| ch | Operator specification: ch - name qubit: target_qubit - index qubit: 1 |
| cnot | Operator specification: cnot - name qubit: control_qubit - index qubit: 0 |
| cnot | Operator specification: cnot - name qubit: control_qubit - index qubit: 1 |
| cnot | Operator specification: cnot - name qubit: target_qubit - index qubit: 0 |
| cnot | Operator specification: cnot - name qubit: target_qubit - index qubit: 1 |
| cp | Operator specification: cp - name qubit: control_qubit - index qubit: 1 |
| cp | Operator specification: cp - name qubit: control_qubit - index qubit: 2 |
| cp | Operator specification: cp - name qubit: target_qubit - index qubit: 1 |
| cp | Operator specification: cp - name qubit: target_qubit - index qubit: 2 |
| crx | Operator specification: crx - name qubit: control_qubit - index qubit: 1 |
| crx | Operator specification: crx - name qubit: control_qubit - index qubit: 2 |
| crx | Operator specification: crx - name qubit: target_qubit - index qubit: 1 |
| crx | Operator specification: crx - name qubit: target_qubit - index qubit: 2 |
| cry | Operator specification: cry - name qubit: control_qubit - index qubit: 1 |
| cry | Operator specification: cry - name qubit: control_qubit - index qubit: 2 |
| cry | Operator specification: cry - name qubit: target_qubit - index qubit: 1 |
| cry | Operator specification: cry - name qubit: target_qubit - index qubit: 2 |
| crz | Operator specification: crz - name qubit: control_qubit - index qubit: 1 |
| crz | Operator specification: crz - name qubit: control_qubit - index qubit: 2 |
| crz | Operator specification: crz - name qubit: target_qubit - index qubit: 1 |
| crz | Operator specification: crz - name qubit: target_qubit - index qubit: 2 |
| cs | Operator specification: cs - name qubit: control_qubit - index qubit: 0 |
| cs | Operator specification: cs - name qubit: control_qubit - index qubit: 1 |
| cs | Operator specification: cs - name qubit: target_qubit - index qubit: 0 |
| cs | Operator specification: cs - name qubit: target_qubit - index qubit: 1 |
| csdg | Operator specification: csdg - name qubit: control_qubit - index qubit: 0 |
| csdg | Operator specification: csdg - name qubit: control_qubit - index qubit: 1 |
| csdg | Operator specification: csdg - name qubit: target_qubit - index qubit: 0 |
| csdg | Operator specification: csdg - name qubit: target_qubit - index qubit: 1 |
| cswap | Operator specification: cswap - name qubit: control_qubit - index qubit: 0 |
| cswap | Operator specification: cswap - name qubit: control_qubit - index qubit: 1 |
| cswap | Operator specification: cswap - name qubit: control_qubit - index qubit: 2 |
| cswap | Operator specification: cswap - name qubit: target_qubit1 - index qubit: 0 |
| cswap | Operator specification: cswap - name qubit: target_qubit1 - index qubit: 1 |
| cswap | Operator specification: cswap - name qubit: target_qubit1 - index qubit: 2 |
| cswap | Operator specification: cswap - name qubit: target_qubit2 - index qubit: 0 |
| cswap | Operator specification: cswap - name qubit: target_qubit2 - index qubit: 1 |
| cswap | Operator specification: cswap - name qubit: target_qubit2 - index qubit: 2 |
| csx | Operator specification: csx - name qubit: control_qubit - index qubit: 0 |
| csx | Operator specification: csx - name qubit: control_qubit - index qubit: 1 |
| csx | Operator specification: csx - name qubit: target_qubit - index qubit: 0 |
| csx | Operator specification: csx - name qubit: target_qubit - index qubit: 1 |
| cu | Operator specification: cu - name qubit: control_qubit - index qubit: 4 |
| cu | Operator specification: cu - name qubit: control_qubit - index qubit: 5 |
| cu | Operator specification: cu - name qubit: target_qubit - index qubit: 4 |
| cu | Operator specification: cu - name qubit: target_qubit - index qubit: 5 |
| cu1 | Operator specification: cu1 - name qubit: control_qubit - index qubit: 1 |
| cu1 | Operator specification: cu1 - name qubit: control_qubit - index qubit: 2 |
| cu1 | Operator specification: cu1 - name qubit: target_qubit - index qubit: 1 |
| cu1 | Operator specification: cu1 - name qubit: target_qubit - index qubit: 2 |
| cu3 | Operator specification: cu3 - name qubit: control_qubit - index qubit: 3 |
| cu3 | Operator specification: cu3 - name qubit: control_qubit - index qubit: 4 |
| cu3 | Operator specification: cu3 - name qubit: target_qubit - index qubit: 3 |
| cu3 | Operator specification: cu3 - name qubit: target_qubit - index qubit: 4 |
| cx | Operator specification: cx - name qubit: control_qubit - index qubit: 0 |
| cx | Operator specification: cx - name qubit: control_qubit - index qubit: 1 |
| cx | Operator specification: cx - name qubit: target_qubit - index qubit: 0 |
| cx | Operator specification: cx - name qubit: target_qubit - index qubit: 1 |
| cy | Operator specification: cy - name qubit: control_qubit - index qubit: 0 |
| cy | Operator specification: cy - name qubit: control_qubit - index qubit: 1 |
| cy | Operator specification: cy - name qubit: target_qubit - index qubit: 0 |
| cy | Operator specification: cy - name qubit: target_qubit - index qubit: 1 |
| cz | Operator specification: cz - name qubit: control_qubit - index qubit: 0 |
| cz | Operator specification: cz - name qubit: control_qubit - index qubit: 1 |
| cz | Operator specification: cz - name qubit: target_qubit - index qubit: 0 |
| cz | Operator specification: cz - name qubit: target_qubit - index qubit: 1 |
| fredkin | Operator specification: fredkin - name qubit: control_qubit - index qubit: 0 |
| fredkin | Operator specification: fredkin - name qubit: control_qubit - index qubit: 1 |
| fredkin | Operator specification: fredkin - name qubit: control_qubit - index qubit: 2 |
| fredkin | Operator specification: fredkin - name qubit: target_qubit1 - index qubit: 0 |
| fredkin | Operator specification: fredkin - name qubit: target_qubit1 - index qubit: 1 |
| fredkin | Operator specification: fredkin - name qubit: target_qubit1 - index qubit: 2 |
| fredkin | Operator specification: fredkin - name qubit: target_qubit2 - index qubit: 0 |
| fredkin | Operator specification: fredkin - name qubit: target_qubit2 - index qubit: 1 |
| fredkin | Operator specification: fredkin - name qubit: target_qubit2 - index qubit: 2 |
| h | Operator specification: h - name qubit: qubit - index qubit: 0 |
| i | Operator specification: i - name qubit: qubit - index qubit: 0 |
| iSwapGate | Operator specification: iSwapGate - name qubit: qubit1 - index qubit: 0 |
| iSwapGate | Operator specification: iSwapGate - name qubit: qubit1 - index qubit: 1 |
| iSwapGate | Operator specification: iSwapGate - name qubit: qubit2 - index qubit: 0 |
| iSwapGate | Operator specification: iSwapGate - name qubit: qubit2 - index qubit: 1 |
| id | Operator specification: id - name qubit: qubit - index qubit: 0 |
| iden | Operator specification: iden - name qubit: qubit - index qubit: 0 |
| iswap | Operator specification: iswap - name qubit: qubit1 - index qubit: 0 |
| iswap | Operator specification: iswap - name qubit: qubit1 - index qubit: 1 |
| iswap | Operator specification: iswap - name qubit: qubit2 - index qubit: 0 |
| iswap | Operator specification: iswap - name qubit: qubit2 - index qubit: 1 |
| measure | Operator specification: measure - name qubit: qubit - index qubit: 0 |
| p | Operator specification: p - name qubit: qubit - index qubit: 1 |
| reset | Operator specification: reset - name qubit: qubit - index qubit: 0 |
| rv | Operator specification: rv - name qubit: qubit - index qubit: 3 |
| rx | Operator specification: rx - name qubit: qubit - index qubit: 1 |
| rxx | Operator specification: rxx - name qubit: qubit1 - index qubit: 1 |
| rxx | Operator specification: rxx - name qubit: qubit1 - index qubit: 2 |
| rxx | Operator specification: rxx - name qubit: qubit2 - index qubit: 1 |
| rxx | Operator specification: rxx - name qubit: qubit2 - index qubit: 2 |
| ry | Operator specification: ry - name qubit: qubit - index qubit: 1 |
| ryy | Operator specification: ryy - name qubit: qubit1 - index qubit: 1 |
| ryy | Operator specification: ryy - name qubit: qubit1 - index qubit: 2 |
| ryy | Operator specification: ryy - name qubit: qubit2 - index qubit: 1 |
| ryy | Operator specification: ryy - name qubit: qubit2 - index qubit: 2 |
| rz | Operator specification: rz - name qubit: qubit - index qubit: 1 |
| rzx | Operator specification: rzx - name qubit: qubit1 - index qubit: 1 |
| rzx | Operator specification: rzx - name qubit: qubit1 - index qubit: 2 |
| rzx | Operator specification: rzx - name qubit: qubit2 - index qubit: 1 |
| rzx | Operator specification: rzx - name qubit: qubit2 - index qubit: 2 |
| rzz | Operator specification: rzz - name qubit: qubit1 - index qubit: 1 |
| rzz | Operator specification: rzz - name qubit: qubit1 - index qubit: 2 |
| rzz | Operator specification: rzz - name qubit: qubit2 - index qubit: 1 |
| rzz | Operator specification: rzz - name qubit: qubit2 - index qubit: 2 |
| s | Operator specification: s - name qubit: qubit - index qubit: 0 |
| sdg | Operator specification: sdg - name qubit: qubit - index qubit: 0 |
| swap | Operator specification: swap - name qubit: qubit1 - index qubit: 0 |
| swap | Operator specification: swap - name qubit: qubit1 - index qubit: 1 |
| swap | Operator specification: swap - name qubit: qubit2 - index qubit: 0 |
| swap | Operator specification: swap - name qubit: qubit2 - index qubit: 1 |
| sx | Operator specification: sx - name qubit: qubit - index qubit: 0 |
| t | Operator specification: t - name qubit: qubit - index qubit: 0 |
| tdg | Operator specification: tdg - name qubit: qubit - index qubit: 0 |
| toffoli | Operator specification: toffoli - name qubit: control_qubit1 - index qubit: 0 |
| toffoli | Operator specification: toffoli - name qubit: control_qubit1 - index qubit: 1 |
| toffoli | Operator specification: toffoli - name qubit: control_qubit1 - index qubit: 2 |
| toffoli | Operator specification: toffoli - name qubit: control_qubit2 - index qubit: 0 |
| toffoli | Operator specification: toffoli - name qubit: control_qubit2 - index qubit: 1 |
| toffoli | Operator specification: toffoli - name qubit: control_qubit2 - index qubit: 2 |
| toffoli | Operator specification: toffoli - name qubit: target_qubit - index qubit: 0 |
| toffoli | Operator specification: toffoli - name qubit: target_qubit - index qubit: 1 |
| toffoli | Operator specification: toffoli - name qubit: target_qubit - index qubit: 2 |
| u1 | Operator specification: u1 - name qubit: qubit - index qubit: 1 |
| u2 | Operator specification: u2 - name qubit: qubit - index qubit: 2 |
| u3 | Operator specification: u3 - name qubit: qubit - index qubit: 3 |
| unitary | Operator specification: unitary - name qubit: qubits - index qubit: 1 |
| x | Operator specification: x - name qubit: qubit - index qubit: 0 |
| y | Operator specification: y - name qubit: qubit - index qubit: 0 |
| z | Operator specification: z - name qubit: qubit - index qubit: 0 |
